src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/oops/InstanceKlass.java

```diff
@@ -111,6 +104,8 @@ private static synchronized void initialize(TypeDataBase db) throws WrongTypeExc
     CLASS_STATE_BEING_INITIALIZED = db.lookupIntConstant("InstanceKlass::being_initialized").intValue();
     CLASS_STATE_FULLY_INITIALIZED = db.lookupIntConstant("InstanceKlass::fully_initialized").intValue();
     CLASS_STATE_INITIALIZATION_ERROR = db.lookupIntConstant("InstanceKlass::initialization_error").intValue();
+    // We need a new fieldsCache each time we attach.
+    fieldsCache = new HashMap<Address, Field[]>();
```

Comment:
This should probably be a WeakHashMap. I tried it and it seems to work (or at least didn't cause any problems). However, when doing a heap dump I didn't notice the table being any smaller on exit when it was made weak, even though there were numerous GC's while dumping the heap.

The <key> is the Address of the hotspot InstanceKlass instance, and this Address is referenced by the SA InstanceKlass mirror. So theoretically when the reference to the mirror goes way, then the cache entry can be cleared.

Reply:
I've changed the map to a WeakHashMap and didn't see any issue during testing. But I didn't measure footprint.